Sources/LanguageServerProtocol/SupportTypes/ServerCapabilities.swift

```diff
@@ -813,6 +903,77 @@ public struct SemanticTokensOptions: WorkDoneProgressOptions, Codable, Hashable
     self.full = full
     self.workDoneProgress = workDoneProgress
   }
+
+  public init?(fromLSPDictionary dictionary: [String : LSPAny]) {
+    guard case .dictionary(let dict) = dictionary["legend"],
+          let legend = SemanticTokensLegend(fromLSPDictionary: dict)
+    else {
+      return nil
+    }
+
+    self.legend = legend
+
+    self.range = nil
+
+    if case .bool(let value) = dictionary["range"] {
+      self.range = .bool(value)
+    }
+
+    if case .dictionary(let dict) = dictionary["range"],
+       let value = SemanticTokensRangeOptions(fromLSPDictionary: dict)
+    {
+      self.range = .value(value)
+    }
+
+    self.full = nil
+    if case .bool(let value) = dictionary["full"] {
+      self.full = .bool(value)
+    }
+
+    if case .dictionary(let dict) = dictionary["range"],
+       let value = SemanticTokensFullOptions(fromLSPDictionary: dict)
+    {
+      self.full = .value(value)
+    }
+
+    if case .bool(let value) = dictionary["workDoneProgress"] {
+      self.workDoneProgress = value
+    }
+  }
+
+  public func encodeToLSPAny() -> LSPAny {
+    var dict: [String: LSPAny] = [:]
+
+    dict["legend"] = legend.encodeToLSPAny()
+
+    if let range {
+      let encodedRange: LSPAny
+      switch range {
+      case .bool(let value):
+        encodedRange = .bool(value)
+      case .value(let rangeOptions):
+        encodedRange = rangeOptions.encodeToLSPAny()
+      }
+      dict["range"] = encodedRange
```

Comment:
We could use a switch expression here, I think

```suggestion
      dict["range"] =  switch range {
      case .bool(let value):
        .bool(value)
      case .value(let rangeOptions):
        rangeOptions.encodeToLSPAny()
      }
```

Reply:
you're so fancy with this new swift :)

Reply:
This is the project where we can actually use all of it because we don’t need to be compatible with older Swift versions ✨